### PR TITLE
Support for decimal literal and null.

### PR DESCRIPTION
**Pull request by praveenbingo**

### CODE REVIEW

File: cpp/src/gandiva/decimal_type_util.h [hunk 1]
```diff
@@ -21,13 +21,37 @@
...
 #include <memory>
+#include <string>
 
 #include "arrow/type.h"
+#include "arrow/util/decimal.h"
 #include "gandiva/arrow.h"
 
 namespace gandiva {
 
+struct DecimalLiteral {
```
Comment (pravindra) [inline cpp/src/gandiva/decimal_type_util.h]: - can we make this a class, and move it to it's own file ? 
- also, name it Decimal128Full ? We'll want to use it beyond literals - for eg. in tests

File: cpp/src/gandiva/decimal_type_util.h [hunk 2]
```diff
@@ -21,13 +21,37 @@
...
+    return precision == other.precision && scale == other.scale && value == other.value;
+  }
+
+  std::string ToString() const {
+    return "(" + value.ToString(0) + "," + std::to_string(precision) + "," +
+           std::to_string(scale) + ")";
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const DecimalLiteral& dec) {
+    os << "(" << dec.value.ToString(0) << "," << std::to_string(dec.precision) << ","
```
Comment (pravindra) [inline cpp/src/gandiva/decimal_type_util.h]: os << dec.ToString() ?

File: cpp/src/gandiva/llvm_generator.cc [hunk 3]
```diff
@@ -628,6 +628,17 @@ void LLVMGenerator::Visitor::Visit(const LiteralDex& dex) {
       value = types->i64_constant(boost::get<int64_t>(dex.holder()));
       break;
 
+    case arrow::Type::DECIMAL: {
+      // build code for struct
+      auto decimal_value = boost::get<DecimalLiteral>(dex.holder());
+      auto decimal_value_in_string = decimal_value.value.ToString(0);
```
Comment (pravindra) [inline cpp/src/gandiva/llvm_generator.cc]: ToString(10) ?

File: java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorDecimalTest.java [hunk 4]
```diff
@@ -43,58 +44,110 @@
...
-        assertFalse(outVector.isNull(i));
-        assertTrue(expOutput[i].compareTo(outVector.getObject(i)) == 0);
-      }
-
-      // free buffers
-      releaseRecordBatch(batch);
-      releaseValueVectors(output);
-      eval.close();
+    int precision = 2;
+    int scale = 0;
```
Comment (pravindra) [inline java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorDecimalTest.java]: why ? let's keep it 38, 8

File: cpp/src/gandiva/proto/Types.proto [hunk 5]
```diff
@@ -146,6 +146,13 @@ message BinaryNode {
   optional bytes value = 1;
 }
 
+message DecimalNode {
+  optional bytes value = 1;
```
Comment (pravindra) [inline cpp/src/gandiva/proto/Types.proto]: string ?

**Comment by praveenbingo** [comment]
Fixed the comments and updated the PR.

File: cpp/src/gandiva/decimal_full.h [hunk 6]
```diff
@@ -55,6 +55,12 @@ class Decimal128Full {
 
   const arrow::Decimal128& value() const { return value_; }
 
+  friend std::ostream& operator<<(std::ostream& os, const Decimal128Full& dec) {
+    os << dec.value().ToString(0) << "," << std::to_string(dec.precision()) << ","
```
Comment (pravindra) [inline cpp/src/gandiva/decimal_full.h]: can we do os << dec.ToString() ?

File: cpp/src/gandiva/decimal_type_sql.h [hunk 7]
```diff
@@ -21,9 +21,10 @@
 #define GANDIVA_DECIMAL_TYPE_SQL_H
 
 #include <algorithm>
+#include <iostream>
```
Comment (pravindra) [inline cpp/src/gandiva/decimal_type_sql.h]: should this be in decimal_full.h ?

File: java/gandiva/src/main/java/org/apache/arrow/gandiva/expression/DecimalNode.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import java.nio.charset.Charset;
+
+import org.apache.arrow.gandiva.exceptions.GandivaException;
+import org.apache.arrow.gandiva.ipc.GandivaTypes;
+
+import com.google.protobuf.ByteString;
+
+
+/**
+ * Used to represent expression tree nodes representing double constants.
```
Comment (pravindra) [inline java/gandiva/src/main/java/org/apache/arrow/gandiva/expression/DecimalNode.java]: decimal constants ?

**Review by pravindra — APPROVED** [review]

**Pull request merged**